### PR TITLE
Puppet Master initial version

### PR DESCRIPTION
**Pull request by Tomatosoup97**

### CODE REVIEW

File: puppetmaster/form_filling.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        try:
+            datum = data[input_name]
+        except KeyError:
+            raise MissingDataException(f"Form is missing '{input_name}' "
+                                       f"input in {data}")
+
+        if type(datum) == tuple:
+            el_type, value = datum  # type: ignore
+        # TODO: this regex should be more flexible / defined in settings
+        elif re.match(r"[0-9]{4}-[0-9]{2}-[0-9]{2}", value) is not None:
```
Comment (wekt0r) [inline puppetmaster/form_filling.py]: Maybe `datetime.strptime(value, "%Y-%m-%d")` or python-dateutil's `parser.parse` would be helpful?

File: puppetmaster/core.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+class SeleniumLiveServerTestCase(SeleniumTestsMixin, StaticLiveServerTestCase):
+    @property
+    def server_url(self):
+        return self.live_server_url
+
+    def create_user(self) -> AbstractUser:
+        return get_user_model().objects.create_user(
+            username='brucelee',
+            password='be_like_water',
+            email='example@gmail.com',
```
Comment (wekt0r) [inline puppetmaster/core.py:111]: Is there any chance that some mails would be sent there? Perhaps "email@example.com" or something similar might be better

File: puppetmaster/core.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        raise NotImplementedError
+
+    @staticmethod
+    def create_driver() -> webdriver.Remote:
+        driver = settings.PM__SELENIUM_DRIVER(
+            executable_path=settings.PM__SELENIUM_DRIVER_PATH)
+        # XXX: Window size must be explicitly set to avoid issue where elements
+        # found in other drivers (e.g. firefox) are not visible in phantomjs.
+        # For more details see: https://github.com/ariya/phantomjs/issues/11637
+        driver.set_window_size(width=1280, height=1024)
```
Comment (wekt0r) [inline puppetmaster/core.py]: extract magic constants

File: puppetmaster/core.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        return self.live_server_url
+
+    def create_user(self) -> AbstractUser:
+        return get_user_model().objects.create_user(
+            username='brucelee',
+            password='be_like_water',
+            email='example@gmail.com',
+            is_staff=False, is_active=True)
+
+    def login_automatically(self):
```
Comment (wekt0r) [inline puppetmaster/core.py]: missing return type

File: puppetmaster/core.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        'django.request': {
+            'handlers': ['console'],
+            'propagate': True,
+            'level': 'DEBUG',
+        },
+    }
+})
+class SeleniumLiveServerTestCase(SeleniumTestsMixin, StaticLiveServerTestCase):
+    @property
+    def server_url(self):
```
Comment (wekt0r) [inline puppetmaster/core.py:104]: missing return type

File: puppetmaster/form_filling.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    def iterate_react_select_options(self, select: WebElement) -> t.List[WebElement]:
+        """Given a .Select rendered by react-select, return an iterable
+        of .Select-option that can be .click()-ed to select them.
+        """
+        # related: https://github.com/JedWatson/react-select/issues/603#issuecomment-157888562 # noqa
+        select.find_element_by_class_name('Select-arrow').click()
+        self.wait(EC.element_to_be_clickable((By.CLASS_NAME, 'Select-option')))
+        return select.find_elements_by_class_name('Select-option')
+
+    def react_select_option(self, select: WebElement, **data):
```
Comment (wekt0r) [inline puppetmaster/form_filling.py]: missing return type and type for data

File: puppetmaster/form_filling.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            .find_element_by_css_selector('input[type="hidden"]') \
+            .get_attribute('name')
+        for option in self.iterate_react_select_options(select):
+            if option.text == data[select_name]:
+                option.click()
+                break
+        else:
+            raise MissingDataException('Option not found.')
+
+    def fill_checkbox(self, input_el: WebElement, values):
```
Comment (wekt0r) [inline puppetmaster/form_filling.py]: missing return type and type for values

File: puppetmaster/form_filling.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            self.get_input_name,
+            form.find_elements_by_css_selector(inputs_selector))
+
+    def get_form_textareas(self, form: WebElement) -> t.Iterable[WebElement]:
+        selector = 'textarea:not(.hidden):not([type="hidden"])'
+        return filter(
+            self.get_input_name,
+            form.find_elements_by_css_selector(selector))
+
+    def fill_form(self, form: WebElement, **data: InputValue):
```
Comment (wekt0r) [inline puppetmaster/form_filling.py]: missing return type

File: puppetmaster/form_filling.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            self.fill_input(form_input, **data)
+
+        for textarea in self.get_form_textareas(form):
+            self.fill_input(textarea, **data)
+
+    def submit_form(self, form: WebElement) -> None:
+        submit_selector = 'button[type="submit"], input[type="submit"]'
+        form.find_element_by_css_selector(submit_selector).click()
+
+    def fill_form_and_submit(self,
```
Comment (wekt0r) [inline puppetmaster/form_filling.py:174]: missing return type

File: puppetmaster/kobotoolbox/assertions.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+from selenium.common.exceptions import TimeoutException
+
+from puppetmaster.exceptions import SeleniumAssertionError
+from puppetmaster.waiter import WaiterInterface
+
+
+class KobotoolboxAssertionsMixin(WaiterInterface):
+    def assert_in_enketo_notification(self, message: str, timeout: int = 1) -> None:
```
Comment (wekt0r) [inline puppetmaster/kobotoolbox/assertions.py:8]: Extract timeout magic constant value

File: puppetmaster/kobotoolbox/assertions.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        selector = '#feedback-bar p'
+        try:
+            self.wait(lambda driver: (
+                message in driver.find_element_by_css_selector(selector).text
+            ), timeout=timeout)
+        except TimeoutException:
+            raise SeleniumAssertionError(f'"{message}" not found in notifications.')
+
+    def assert_in_kpi_notifications(self, message: str, timeout: int = 1) -> None:
```
Comment (wekt0r) [inline puppetmaster/kobotoolbox/assertions.py:17]: Extract timeout magic constant value

File: puppetmaster/kobotoolbox/form_filling.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        try:
+            actual_date_input.clear()
+        except InvalidElementStateException:
+            pass
+        actual_date_input.send_keys(date)
+        # XXX: After inputting the data a datepicker widget appears, so we need
+        # to hide it by clicking on the blank area nearby
+        ActionChains(self.driver)\
+            .move_to_element(actual_date_input)\
+            .move_by_offset(-300, 0).click().perform()
```
Comment (wekt0r) [inline puppetmaster/kobotoolbox/form_filling.py]: extract magic constant

File: puppetmaster/waiter.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            ).until(condition)
+        except TimeoutException as e:
+            # Take screenshot before failure and explicitly re-raise exception
+            self.make_screenshot()
+            raise e
+
+    def wait(self,
+             condition: t.Callable[[webdriver.Remote], t.Any],
+             timeout: int = 1,
+             max_tries: int = 3) -> WebDriverWait:
```
Comment (wekt0r) [inline puppetmaster/waiter.py]: extract magic constant

**Review by wekt0r — CHANGES_REQUESTED** [review]

File: puppetmaster/core.py (hunk 2 above)
Comment (Tomatosoup97) [reply puppetmaster/core.py:111]: example@gmail.com is just as exemplary as email@example.com -> it's not allowed to reserve that address under gmail domain

File: puppetmaster/core.py (hunk 5 above)
Comment (Tomatosoup97) [reply puppetmaster/core.py:104]: Done

File: puppetmaster/core.py (hunk 3 above)
Comment (Tomatosoup97) [reply puppetmaster/core.py]: Done

File: puppetmaster/core.py (hunk 4 above)
Comment (Tomatosoup97) [reply puppetmaster/core.py]: Done

File: puppetmaster/form_filling.py (hunk 6 above)
Comment (Tomatosoup97) [reply puppetmaster/form_filling.py]: Done

File: puppetmaster/form_filling.py (hunk 7 above)
Comment (Tomatosoup97) [reply puppetmaster/form_filling.py]: Done

File: puppetmaster/form_filling.py (hunk 8 above)
Comment (Tomatosoup97) [reply puppetmaster/form_filling.py]: Done

File: puppetmaster/form_filling.py (hunk 9 above)
Comment (Tomatosoup97) [reply puppetmaster/form_filling.py:174]: Done

File: puppetmaster/kobotoolbox/assertions.py (hunk 10 above)
Comment (Tomatosoup97) [reply puppetmaster/kobotoolbox/assertions.py:8]: I wouldn't name it a magic constant
- it's pretty clear what it means, just a default timeout set to 1 
- In selenium you want to have small timeouts, like one second, and then explicitly make it longer in places where something takes time to load, so it's by design not recommended to extend that default time
- putting `DEFAULT_TIMEOUT` into signatures would just make them extend pep line length limit without making it more readable

File: puppetmaster/kobotoolbox/form_filling.py (hunk 12 above)
Comment (Tomatosoup97) [reply puppetmaster/kobotoolbox/form_filling.py]: Done

File: puppetmaster/waiter.py (hunk 13 above)
Comment (Tomatosoup97) [reply puppetmaster/waiter.py]: Done

File: puppetmaster/form_filling.py (hunk 1 above)
Comment (Tomatosoup97) [reply puppetmaster/form_filling.py]: Good catch, done.

**Review by wekt0r — APPROVED** [review]

**Pull request merged**